pkg/queue/cluster_queue_impl.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"container/heap"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	kueue "sigs.k8s.io/kueue/api/v1alpha1"
+	"sigs.k8s.io/kueue/pkg/workload"
+)
+
+// ClusterQueueImpl is the base implementation of ClusterQueue interface.
+// It can be inherited and overwritten by other class.
+type ClusterQueueImpl struct {
+	// QueueingStrategy indicates the queueing strategy of the workloads
+	// across the queues in this ClusterQueue.
+	QueueingStrategy kueue.QueueingStrategy
+
+	heap heapImpl
+}
+
+var _ ClusterQueue = &ClusterQueueImpl{}
+
+func (c *ClusterQueueImpl) Update(apiCQ *kueue.ClusterQueue) {
+	c.QueueingStrategy = apiCQ.Spec.QueueingStrategy
+}
+
+func (c *ClusterQueueImpl) AddFromQueue(q *Queue) bool {
+	added := false
+	for _, w := range q.items {
+		if c.PushIfNotPresent(w) {
+			added = true
+		}
+	}
+	return added
+}
+
+func (c *ClusterQueueImpl) DeleteFromQueue(q *Queue) {
+	for _, w := range q.items {
+		c.Delete(w.Obj)
+	}
+}
+
+func (c *ClusterQueueImpl) PushIfNotPresent(info *workload.Info) bool {
+	item := c.heap.items[workload.Key(info.Obj)]
+	if item != nil {
+		return false
+	}
+	heap.Push(&c.heap, *info)
+	return true
+}
+
+func (c *ClusterQueueImpl) PushOrUpdate(w *kueue.QueuedWorkload) {
+	item := c.heap.items[workload.Key(w)]
+	info := *workload.NewInfo(w)
+	if item == nil {
+		heap.Push(&c.heap, info)
+		return
+	}
+	item.obj = info
+	heap.Fix(&c.heap, item.index)
+}
+
+func (c *ClusterQueueImpl) Delete(w *kueue.QueuedWorkload) {
+	item := c.heap.items[workload.Key(w)]
+	if item != nil {
+		heap.Remove(&c.heap, item.index)
+	}
+}
+
+func (c *ClusterQueueImpl) Pop() *workload.Info {
+	if c.heap.Len() == 0 {
+		return nil
+	}
+	w := heap.Pop(&c.heap).(workload.Info)
+	return &w
+}
+
+func (c *ClusterQueueImpl) Pending() int32 {
+	return int32(len(c.heap.heap))
+}
+
+func (c *ClusterQueueImpl) RequeueWorkload(wInfo *workload.Info) (bool, error) {
```

Comment:
Why do we have this function instead of just `PushIfNotPresent`?

I suspect it has to do with the new policies, but I better review that in the other PR, so I can see the big picture.

Reply:
sg. I delete this func. Will add it in the other PR.